runtimes/rust/lbr-prelude/src/json.rs

```diff
@@ -0,0 +1,327 @@
+            Err(val) => Ok(sum_constructor("Left", vec![val.to_json()?])),
+        }
+    }
+
+    fn from_json(value: Value) -> Result<Self, Error> {
+        let map = sum_parser(&value)?;
+
+        match map {
+            ("Right", ctor_fields) => 
+                match &ctor_fields[..] {
+                    [val] => Ok(Ok(T::from_json(val.clone())?)),
+                    _ => error(&format!("Expected a JSON Array with 1 fields but got {:?}", ctor_fields)),
+                },
+                ("Left", ctor_fields) => 
+                    match &ctor_fields[..] {
+                        [val] => Ok(Err(E::from_json(val.clone())?)),
+                        _ =>  error(&format!(
+                                        "Expected a JSON Array with 1 fields but got {:?}",
+                                        ctor_fields
+                                        )),
+                    },
+                _ => error(
+                        "Expected a JSON String to contain one of constructor names (Just, Nothing) but got an unknown constructor name.")
+        }
+    }
+}
+
+impl<T> Json for Vec<T>
+where
+    T: Json,
+{
+    fn to_json(&self) -> Result<Value, Error> {
+        let values = self
+            .iter()
+            .map(|val| val.to_json())
+            .into_iter()
+            .collect::<Result<Vec<Value>, Error>>()?;
+
+        Ok(Value::Array(values))
+    }
+
+    fn from_json(value: Value) -> Result<Self, Error> {
+        match value {
+            Value::Array(vec) => vec
+                .iter()
+                .map(|val| T::from_json(val.clone()))
+                .into_iter()
+                .collect::<Result<Vec<T>, Error>>(),
+            _ => error(&format!("Expected a JSON Array but got {:?}", value)),
+        }
+    }
+}
+
+impl<T> Json for HashSet<T>
+where
+    T: Json + Eq + Hash,
+{
+    fn to_json(&self) -> Result<Value, Error> {
+        let values = self
+            .iter()
+            .map(|val| val.to_json())
+            .into_iter()
+            .collect::<Result<Vec<Value>, Error>>()?;
+
+        Ok(Value::Array(values))
+    }
+
+    fn from_json(value: Value) -> Result<Self, Error> {
+        match value {
+            Value::Array(vec) => {
+                let set = vec
+                    .iter()
+                    .map(|val| T::from_json(val.clone()))
+                    .into_iter()
+                    .collect::<Result<HashSet<T>, Error>>()?;
+
+                if set.len() == vec.len() {
+                    Ok(set)
+                } else {
+                    error(&format!(
+                        "Expected the JSON Array to have all unique elements: {:?}",
+                        vec
+                    ))
+                }
+            }
+            _ => error(&format!("Expected a JSON Array but got {:?}", value)),
+        }
+    }
+}
+
+impl<K, V> Json for HashMap<K, V>
+where
+    K: Json + Eq + Hash,
```

Comment:
hmm, but in that case we would have to convert to a Value first, which feels like a bit of an unnecessary overhead.

Reply:
Looks to me like key.to_json() is already doing that no?

Reply:
a Hash trait implementation is required for the key, when we build the HashMap of the deserialized Map, so the original key must be hashable.
This would also change to an `Ord` instead of `Hash` if we choose `BTreeMap`.

Reply:
Leaving this for you to decide, we can easily adjust if needed.